ideascube/serveradmin/catalog.py

```diff
@@ -0,0 +1,459 @@
+    pass
+
+
+class InvalidPackageType(Exception):
+    pass
+
+
+class InvalidHandlerType(Exception):
+    pass
+
+
+class NoSuchPackage(Exception):
+    pass
+
+
+class Remote:
+    def __init__(self, id, name, url):
+        self.id = id
+        self.name = name
+        self.url = url
+
+    @classmethod
+    def from_file(cls, path):
+        with open(path, 'r') as f:
+            d = yaml.safe_load(f.read())
+
+            try:
+                return cls(d['id'], d['name'], d['url'])
+
+            except KeyError as e:
+                raise InvalidFile(
+                    'Remote file is missing a {} key: {}'.format(e, path))
+
+    def to_file(self, path):
+        with open(path, 'w') as f:
+            d = {'id': self.id, 'name': self.name, 'url': self.url}
+            f.write(yaml.safe_dump(d, default_flow_style=False))
+
+
+class Package:
+    def __new__(cls, id, metadata):
+        try:
+            cls = cls.__TYPES__[metadata['type']]
+
+        except KeyError:
+            raise InvalidPackageType(metadata['type'])
+
+        pkg = object.__new__(cls)
+        pkg.__init__(id, metadata)
+
+        return pkg
+
+    def __init__(self, id, metadata):
+        self.id = id
+        self._metadata = metadata
+
+    def __getattr__(self, name):
+        try:
+            return self._metadata[name]
+
+        except KeyError:
+            raise AttributeError(name)
+
+    @property
+    def version(self):
+        # 0 has the advantage of always being "smaller" than any other version
+        return self._metadata.get('version', '0')
+
+    def install(self, download_path, install_dir):
+        raise NotImplementedError('Subclasses must implement this method')
+
+    def remove(self, install_dir):
+        raise NotImplementedError('Subclasses must implement this method')
+
+
+class ZippedZim(Package):
+    def install(self, download_path, install_dir):
+        if not zipfile.is_zipfile(download_path):
+            os.unlink(download_path)
+            raise InvalidFile('{} is not a zip file'.format(download_path))
+
+        with zipfile.ZipFile(download_path, "r") as z:
+            names = z.namelist()
+            names = filter(lambda n: n.startswith('data/'), names)
+            z.extractall(install_dir, members=names)
+
+        zimname = os.path.basename(self.url).split('+', 1)[-1][:-4] + '.zim'
+        datadir = os.path.join(install_dir, 'data')
+
+        for zim in glob(os.path.join(datadir, '*', '{}*'.format(zimname))):
+            new = zim.replace(zimname, '{0.id}.zim'.format(self))
+            os.rename(zim, new)
+
+    def remove(self, install_dir):
+        zimname = '{0.id}.zim*'.format(self)
+        datadir = os.path.join(install_dir, 'data')
+
+        for path in glob(os.path.join(datadir, '*', zimname)):
+            try:
+                os.unlink(path)
+
+            except IsADirectoryError:
+                shutil.rmtree(path)
+
+
+# Register the package types
+Package.__TYPES__ = {
```

Comment:
Dunno, this felt simple and elegant, since the `Catalog` can just always use the `Package()` constructor, and `Package` itself figures out which package type to return.

How would you do the registry with a metaclass?